### ISSUE TITLE
DEALLOCATE: syntax error at end of input at character 12

### ISSUE THREAD

**Issue by zeha** [PowerDNS/pdns#4201]
This is Debian bug#831741, about pdns-backend-pgsql:

```
The pg log gets spammed with:

2016-07-18 20:45:26 GMT ERROR:  syntax error at end of input at character 12
2016-07-18 20:45:26 GMT STATEMENT:  DEALLOCATE

constantly, hundreds of times per second.

I had to set log_min_messages and log_min_error_statement to fatal to avoid it.

Something is wrong with the sql pdns sends to pg.

This is a recent regression.  I did not notice with with the first 4.0 pres in sid.
```

(And indeed, the line adding the `DEALLOCATE` statement was added in 4.0.0-rc1.)

**Comment by jhcloos** [comment]
In my (debian bug reporter) case, my pdns is a hidden master serving a number of nsds via ?XFR.  All pdns was logging was AXFRs.

The number of AXFRs is high lately; either a bug in nsd or nsd doesn't like power 4's axfr syntax?

**Comment by rgacogne** [comment]
Alright, we use a new instance of `UeberBackend` to serve AXFRs, creating and destroying a `gPgSQLBackend` after using very few statements. All the unused ones will have an empty `d_stmt`, causing this error.

**Comment by rgacogne** [comment]
@jhcloos regarding the high number of AXFRs, do you have some metrics you would be willing to share? Anything in NSD's logs?

**Comment by jhcloos** [comment]
Nsd isn't logging anything; I was going to see about making it more
verbose, but did not have time today to do so.

I've done that now on one of the slaves.

On restart it did an axfr of everything, as normal.

Since then it is not logging anything.

-JimC
-- 
James Cloos <cloos@jhcloos.com>         OpenPGP: 0x997A9F17ED7DAEA6

**Comment by Habbie** [comment]
@jhcloos are you able to test #4206?

**Comment by zeha** [comment]
Patched pdns packages (4.0.0-5) are on their way.

**Comment by jhcloos** [comment]
I updated to the new version in sid last night.

The dns servicing kept working fine.

Today I reset the postgres logging from fatal to notice.

And the error messages are gone.

I still don't know why my 2ndaries are making so many axfrs.

(I did not change my nsd configs when power added ixfr support, so that
is why they are all axfr and not ixfr. But beyond that ....)

-JimC
-- 
James Cloos <cloos@jhcloos.com>         OpenPGP: 0x997A9F17ED7DAEA6

**Comment by jhcloos** [comment]
I just saw this in my pg logs:

2016-07-23 07:23:21 GMT ERROR:  prepared statement "stmt146925860186925" already exists
2016-07-23 07:23:21 GMT STATEMENT:  select cryptokeys.id, flags, case when active then 1 else 0 end as active, content from domains, cryptokeys where cryptokeys.domain_id=domains.id and name=$1
2016-07-23 07:23:21 GMT ERROR:  prepared statement "stmt146925860186925" does not exist
2016-07-23 07:23:21 GMT STATEMENT:  DEALLOCATE stmt146925860186925

**Comment by rgacogne** [comment]
Do you have anything related in the pdns logs? At first glance it looks like using only `gettimeofday()` to generate temporary statement names might not be enough to guarantee that they are unique per db session.

**Comment by jhcloos** [comment]
>>>>> "RG" == Remi Gacogne <notifications@github.com> writes:

RG> Do you have anything related in the pdns logs?

Just the still constant stream of excessive axfrs.

I'm working towards switching to sql replication to avoid that (I
presume) nsd bug.

-JimC
-- 
James Cloos <cloos@jhcloos.com>         OpenPGP: 0x997A9F17ED7DAEA6

**Comment by pieterlexis** [comment]
Symptom removal in #4206

**Comment by Habbie** [comment]
I merged #8627, do we agree that we can then close this ticket?

**Comment by zeha** [comment]
I think so!
